### ISSUE TITLE
Socket ReceiveFrom() and SendTo() issues on Linux after .NET 5 upgrade

### ISSUE THREAD

**Issue by ShreyasZare** [dotnet/runtime#45651]
### Description
An open source DNS server project that I am working currently uses .NET 3.1 and works well on Linux and Windows. I upgraded the project to .NET 5 and it stop functioning immediately without any kind of errors being generated. So I started investigating the issue and used Console.WriteLine and tcpdump on capture packets on all networks on the Ubuntu 20.04 VM.

### Configuration

* The code is running .NET 5. Earlier the same code worked perfectly on .NET 3.1.
* Tested on Ubuntu 20.04 LTS x64.
* OS is running in a VM with 1 CPU and 1GB RAM.

### Regression?

From the investigations, it seems there is an issue with how Socket handles ReceiveFrom() and SendTo() in .NET 5 upgrade. The ReceiveFrom() is giving wrong `remoteEP` values after a couple of packets. The remoteEP values keep on repeating in the Console.WriteLine() which meant that the method is giving wrong EndPoint for the received packet.

Further more, the SendTo() method sends the packets to a `remoteEP` different from the one provided in the argument. It too seems to gets stuck on the same end point after sending a couple of packets and then keeps on sending packets to destination that was not specified in the `remoteEP` argument.

This issue comes only on Linux and Windows seems to be not affected.

### Other information

Attaching PCAP files to inspect in a zip file: [dns-issue-pcap.zip](https://github.com/dotnet/runtime/files/5648857/dns-issue-pcap.zip)

1. `dns-issue1.pcap` : In here ICMP port unreachable can be observed as the DNS server responds to a request with the stuck end point "127.0.0.1:36145". Expected destination end point is "192.168.180.1:54290" for the last response in the pcap.

2. `dns-issue2.pcap`: In here the DNS server is sending response to a stuck end point where the IP seems to be correct but the port number is different ie. "192.168.180.1:58002". Expected destination end point is "192.168.180.1:62807" for the last response in the pcap.

The link to the affected DNS server project code to show the .NET 3.1 working implementation that broke with .NET 5 upgrade: https://github.com/TechnitiumSoftware/DnsServer/blob/master/DnsServerCore/Dns/DnsServer.cs#L224-L354

**Comment by scalablecory** [comment]
triaging for investigation in 6.0. CC @antonfirsov @tmds

**Comment by tmds** [comment]
@ShreyasZare can you provide a small reproducer?

**Comment by ShreyasZare** [comment]
Thanks for the replies. I have create a small dns proxy app that can reproduce this issue on .NET 5 on Ubuntu 20.04 LTS.

Here is the link to the repo: [https://github.com/ShreyasZare/DnsProxy](https://github.com/ShreyasZare/DnsProxy)

While I created and tested this app, it was noticed that the code works well when SendTo() is used but actually fails when async call BeginSendTo() is being used.

Dotnet version installed on Ubuntu 20.04 LTS: `Microsoft.NETCore.App 5.0.0 [/opt/dotnet/shared/Microsoft.NETCore.App]`

Steps to reproduce the issue:
1. Copy and run the project output in ubuntu machine with .NET 5.0.0 installed: `dotnet DnsProxy.dll`
2. Use `nslookup www.google.com <ip-address>` to query the app. Where `<ip-address>` is the IP address of the ubuntu machine. If on the same machine use loopback (127.0.0.1) address.
3. You will notice that `nslookup` gives request timeout messages as shown below:
```
C:\Users\user>nslookup www.google.com 192.168.180.129
DNS request timed out.
    timeout was 2 seconds.
Server:  UnKnown
Address:  192.168.180.129

DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
*** Request to UnKnown timed-out
```
4. Repeat the same experiment on Windows with .NET 5 installed and you will notice that `nslookup` returns response as expected.

Observations:
1. The dns proxy app's console output will look as shown below. The end point address that is printed remains same as that of the remote end point of the first UDP packet that was received. The app keeps transmitting responses to this stuck end point that ReceiveFrom() keeps returning for each subsequent packet.
```
DNS proxy is running. Press CTRL+C to terminate.
ReceiveFrom: 192.168.180.1:64816
SendTo: 192.168.180.1:64816
ReceiveFrom: 192.168.180.1:64816
SendTo: 192.168.180.1:64816
ReceiveFrom: 192.168.180.1:64816
SendTo: 192.168.180.1:64816
ReceiveFrom: 192.168.180.1:64816
SendTo: 192.168.180.1:64816
ReceiveFrom: 192.168.180.1:64816
SendTo: 192.168.180.1:64816
ReceiveFrom: 192.168.180.1:64816
SendTo: 192.168.180.1:64816
ReceiveFrom: 192.168.180.1:64816
SendTo: 192.168.180.1:64816
ReceiveFrom: 192.168.180.1:64816
SendTo: 192.168.180.1:64816
```
2. When `nslookup` is tested from the same ubuntu console then the following output is seen. In here the loopback endpoint gets stuck and when `nslookup` query is made from another computer on the network, the app sends UDP responses to the stuck loopback end point. These responses can be seen if you run tcpdump and observe the output which is similar to the pcap that were provided earlier.
```
DNS proxy is running. Press CTRL+C to terminate.
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
ReceiveFrom: 127.0.0.1:56960
SendTo: 127.0.0.1:56960
```

The code in question in marked here to inspect: [https://github.com/ShreyasZare/DnsProxy/blob/master/DnsProxy/DnsProxyService.cs#L54-L62](https://github.com/ShreyasZare/DnsProxy/blob/master/DnsProxy/DnsProxyService.cs#L54-L62)

There are 3 ways tried to send response. You can test each one by commenting the other. The first one that uses SendTo() blocking call works as expected but the other two fail.

The other two calls however work on Windows with .NET 5 and work for both Windows and Linux with .NET Core 3.1.

**Comment by wfurt** [comment]
perhaps related to #39313? cc: @CarnaViire

**Comment by ShreyasZare** [comment]
This issue does not occur when using ReceiveFromAsync() with SendToAsync() and thus updating my code to use ReceiveFromAsync(). 

This change also improves the performance on top of already significant performance improvements from moving to .NET 5 by itself.

**Comment by antonfirsov** [comment]
Removing the caching of the `LocalEndpoint` doesn't help. With Debug builds we end up with an assertion failure in `SocketAsyncContext.Unix`:
https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L475

The assertion keeps failing when switching from `ReceiveFrom` to `ReceiveFromAsync` as suggested in https://github.com/dotnet/runtime/issues/45651#issuecomment-744010117.

Might be a side effect of the .NET 5 performance refactors.

**Comment by tmds** [comment]
@antonfirsov, since you can reproduce the issue, can you check if https://github.com/dotnet/runtime/pull/45683 has an effect?

**Comment by tmds** [comment]
I think we are missing `SetReceivedFlags = true` here:

https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L1510-L1516

**Comment by antonfirsov** [comment]
@tmds yeah `SetReceivedFlags = true` seems to fix the issue. Will raise a PR tomorrow.

**Comment by tmds** [comment]
I wonder why tests haven't caught this.

**Comment by geoffkizer** [comment]
> I wonder why tests haven't caught this.

+1. I assume we just don't have tests for this particular combination (async ReceiveFrom with Memory) or something like that, but it's still concerning.

**Comment by antonfirsov** [comment]
I want to produce a minimal repro and add a new test to the PR. It has probably something to do with `SendTo` & `ReceiveFrom` happening from the same socket which is not exercised by our current UDP tests.

Unfortunately some new (?) infra changes seemingly broke my entire Linux workflow, I'm not sure anymore if I can finish the troubleshooting and do any meaningful progress before my vacation.

**Comment by antonfirsov** [comment]
Reopening until it gets backported.

**Comment by karelz** [comment]
Fixed in 5.0.3 in PR #46629 and in 6.0 (master) in PR #46151
